src/mainwindow.cpp

```diff
@@ -428,21 +428,27 @@ void MainWindow::loadFolder(const Fm::FilePath & newFolderPath) {
 
 // the image is loaded (the method is only called if the loading is not cancelled)
 void MainWindow::onImageLoaded() {
-  image_ = loadJob_->image();
+  // Note: As the signal finished() is emitted from different thread,
+  // we can get it even after canceling the job (and setting the loadJob_
+  // to nullptr). This simple check should be enough.
+  if (sender() == loadJob_)
+  {
+    image_ = loadJob_->image();
```

Comment:
That should help but it's still not really safe, the other thread could definitely change `loadJob_` to `nullptr` between the `if` and this line. And this thread could also set it to `nullptr` between the `nullptr` check and the cancel in the other thread.

Even if you were also using `sender()` on this line I believe the other accesses to `loadJob_` (`==` and `= nullptr`) would still technically be undefined behaviour. We need to either make sure we don't need to access it concurrently or just lock it with a mutex when we do.